File: android/src/main/java/com/josephblough/flutterdrivingdirections/FlutterDrivingDirectionsPlugin.java
package com.josephblough.flutterdrivingdirections;

import android.content.Intent;
import android.net.Uri;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterDrivingDirectionsPlugin */
public class FlutterDrivingDirectionsPlugin implements MethodCallHandler {
  final private Registrar registrar;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_driving_directions");
    channel.setMethodCallHandler(new FlutterDrivingDirectionsPlugin(registrar));
  }

  private FlutterDrivingDirectionsPlugin(Registrar registrar) {
    this.registrar = registrar;
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("launchDirections")) {
      result.notImplemented();
    }

    Double latitude = call.argument("latitude");
    Double longitude = call.argument("longitude");
    String address = call.argument("address");

    String uri = "http://maps.google.com/maps?daddr=" + latitude + "," + longitude + " (" + address + ")";
    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
    intent.setPackage("com.google.android.apps.maps");
    registrar.activity().startActivity(intent);
  }
}

File: android/src/main/kotlin/com/josephblough/flutter_driving_directions/FlutterDrivingDirectionsPlugin.kt
package com.josephblough.flutter_driving_directions

import androidx.annotation.NonNull

import android.content.Intent
import android.net.Uri
import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterDrivingDirectionsPlugin */
class FlutterDrivingDirectionsPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  // Activity needed to launch the intent
  private var activity : Activity? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_driving_directions")
    channel.setMethodCallHandler(this)
  }

  override fun onAttachedToActivity(@NonNull binding: ActivityPluginBinding) {
    activity = binding.getActivity()
  }

  override fun onReattachedToActivityForConfigChanges(@NonNull binding: ActivityPluginBinding) {
    activity = binding.getActivity()
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method != "launchDirections") {
      result.notImplemented()
    }

    val latitude = call.argument<Double>("latitude")
    val longitude = call.argument<Double>("longitude")
    val address = call.argument<String>("address")

    val uri = "http://maps.google.com/maps?daddr=$latitude,$longitude ($address)"
    val intent = Intent(Intent.ACTION_VIEW, Uri.parse(uri))
    intent.setPackage("com.google.android.apps.maps")
    activity?.startActivity(intent)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
